osquery/tables/system/windows/windows_search.cpp

```diff
@@ -0,0 +1,478 @@
+    return "";
+  }
+  auto const pSearchCatalogManagerGuard = scope_guard::create(
+      [pSearchCatalogManager]() { pSearchCatalogManager->Release(); });
+
+  // Call ISearchCatalogManager::GetQueryHelper to get the ISearchQueryHelper
+  // interface
+  ISearchQueryHelper* pQueryHelper = nullptr;
+  hr = pSearchCatalogManager->GetQueryHelper(&pQueryHelper);
+  if (FAILED(hr)) {
+    LOG(ERROR) << windowsSearchTableName << ": failed to get query helper";
+    return "";
+  }
+  auto const pQueryHelperGuard =
+      scope_guard::create([pQueryHelper]() { pQueryHelper->Release(); });
+
+  hr = pQueryHelper->put_QueryMaxResults(maxResults);
+  if (FAILED(hr)) {
+    LOG(ERROR) << windowsSearchTableName << ": failed to set max results";
+    return "";
+  }
+
+  if (!columns.empty()) {
+    // TODO: find a way to verify that the columns requested exist before the
+    // query else we just get an error. If a new column is added or dropped on
+    // an OS could break existing queries.
+    std::string selectColumns;
+    for (const auto& k : columns) {
+      selectColumns += k + ",";
+    }
+
+    // Remove the last comma
+    if (!selectColumns.empty()) {
+      selectColumns.pop_back();
+    }
+
+    hr = pQueryHelper->put_QuerySelectColumns(
+        stringToWstring(selectColumns).c_str());
+    if (FAILED(hr)) {
+      LOG(ERROR) << windowsSearchTableName << ": failed to set columns";
+      return "";
+    }
+  }
+
+  if (!sort.empty()) {
+    hr = pQueryHelper->put_QuerySorting(stringToWstring(sort).c_str());
+    if (FAILED(hr)) {
+      LOG(ERROR) << windowsSearchTableName << ": failed to set sort";
+      return "";
+    }
+  }
+
+  LPWSTR sql;
+  hr = pQueryHelper->GenerateSQLFromUserQuery(
+      stringToWstring(userInput).c_str(), &sql);
+  if (FAILED(hr)) {
+    LOG(ERROR) << windowsSearchTableName
+               << ": failed to generate SQL from user query";
+    return "";
+  }
+
+  std::string ret = wstringToString(sql);
+  CoTaskMemFree(sql);
+  return ret;
+}
+
+QueryData genWindowsSearch(QueryContext& context) {
+  QueryData results;
+  HRESULT hr = CoInitializeEx(0, COINIT_MULTITHREADED);
+  auto const coUninitalizeGuard =
+      scope_guard::create([hr]() { CoUninitialize(); });
+
+  CDataSource cDataSource;
+  hr = cDataSource.OpenFromInitializationString(
+      L"provider=Search.CollatorDSO.1;EXTENDED "
+      L"PROPERTIES=\"Application=Windows\"");
+  if (FAILED(hr)) {
+    LOG(ERROR) << windowsSearchTableName << ": error initializing CDataSource";
+    return results;
+  }
+  auto const cDataSourceGuard =
+      scope_guard::create([&cDataSource]() { cDataSource.Close(); });
+
+  CSession cSession;
+  hr = cSession.Open(cDataSource);
+  if (FAILED(hr)) {
+    LOG(ERROR) << windowsSearchTableName << ": error opening CSession";
+    return results;
+  }
+  auto const cSessionGuard =
+      scope_guard::create([&cSession]() { cSession.Close(); });
+
+  LONG maxResults = 100;
```

Comment:
there is no need to restrict the results to 100 by default. You can set -1 to get all results. See doc [here](https://learn.microsoft.com/en-us/windows/win32/api/searchapi/nf-searchapi-isearchqueryhelper-put_querymaxresults) and example [here](https://github.com/microsoft/Windows-classic-samples/blob/7cbd99ac1d2b4a0beffbaba29ea63d024ceff700/Samples/Win7Samples/winui/WindowsSearch/WSOleDB/WSOleDB.cpp#LL143C42-L143C44).

Reply:
@marcosd4h , this was intentional, because I wanted someone to be able to run `select * from windows_search` and quickly be able to understand the table.

If I run `select * form windows_search` with no default max value ... it will take a very long time to return the results. I think most people will just assume it's not working or crashed or something.

However, I could imagine the frustration someone might feel since it's not obvious max_results is set by default.

Perhaps we should consider not hiding the `max_results` column?

Or maybe we it's okay to let the default query be slow?

Reply:
I think it's reasonable to set the default max results to something. The intent it to allow a simple `select *` and similar exploration. (We've had similar discussion around the various log tables)

Probably note it in the schema, as well as how to unlimit the results.